### DISCUSSION TITLE
Use `HEAD` request to DockerHub to avoid hitting rate limit

### DISCUSSION THREAD

**Discussion by ChrisJBurns** [renovatebot/renovate#21964] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

I would like Renovate to be able to avoid hitting rate limits on DockerHub when it tries to determine whether there are new image versions available. This is how [Harbor avoids hitting](https://goharbor.io/docs/2.1.0/administration/configure-proxy-cache) the DockerHub rate limits with its proxy cache and would be great if Renovate was to do the same to avoid hitting the same limits. This would save the need for having to configure proxy caches.

https://github.com/renovatebot/renovate/issues/11473 seems to suggest this was implemented but I'm not sure it satisfies my case

**Comment by rarkins** [comment]
We can't simply do HEAD instead of every GET. You're welcome to experiment and submit a PR though

**Reply by ChrisJBurns** [reply]
If that's the case, do you know I can avoid the DockerHub rate limits? I've done some reading and I've read that using a Proxy Cache doesn't really seem to solve our problem of knowing when there are new versions of images? Or does Renovate not exhaust a pull when it's checking for new image versions? Because otherwise due to the fact that we have more than a hundred images in a monorepo that we use to build third party images, renovate will be exhausting the entire 100 pull count before it even completes renovating that repository - it's then got many other repositories to do after that.

**Reply by rarkins** [reply]
If you need to query Docker Hub more heavily then their free API limit allows, you should have a paying Docker account and configure a token for it using hostRules

**Reply by ChrisJBurns** [reply]
Don't think that's possible for us in our case. I guess we will have to turn off our Docker functionality in renovate.

**Reply by ChrisJBurns** [reply]
@rarkins  Just so I more deeply understand the way renovate works.

Which one of the following cases are true?:
1) Renovate only ever does a qualifying pull (qualifying pull meaning a pull that counts towards 1 of the free 100 you get per 6 hours on DockerHub) when it checks for newer versions of an image **_and_** it's **_found_** one
2) Exhaust a qualifying pull regardless of whether there are newer images or not on DockerHub. So if I'm running latest version of busybox in a Dockerfile and if it searches for new versions and can't find any on DockerHub, it still quantifies this as a qualifying pull.

I ask this because if it only ever exhausts a qualifying pull in case 1, then this makes our case a lot less problematic, because we've just turned Renovate on for the first time and it's identified a tonne of images that need to be updated, which isn't a worry for now because we're hoping when we get closer to latest versions of all of the images, we exhaust less qualifying pulls per 6 hours as there will be less and less newer images found. 

The reason I ask for distinction is because if it is indeed case 2, this means that even if all of our images where up to date, we'd still hit the limit which means in a repo with hundreds of images, we'd never actually get to a point where we can get renovate to recognise if there are updates without hitting the limit.

Apologies for the hassle, I just really want to understand in what circumstances it would actually do a qualifying pull so I can let folks know in my org that if it's case 1 the chances of us hitting rate limit's will go down as the weeks go by whilst we update our images.

**Reply by rarkins** [reply]
I'm not sure which of Renovate's requests to Docker Hub count towards the API limit. Essentially it queries the list of tags for an image and then for each update it fetched the digest. It does not pull a full image at either time

**Reply by viceice** [reply]
get requests to manifest files count. by default we try to do head request to fetch the digest. but we do one get per image to get the labels from an image to lookup sourceUrl for changelog

**Reply by ChrisJBurns** [reply]
@viceice So knowing that, which cases that I've outlined above do you think would be true?

**Reply by viceice** [reply]
it's a litle bit of both, because renovate can cache responses for runs. also pulling official images or bitnami images are not counted to rate limit

**Reply by viceice** [reply]
you should configure redis cache, so renovate can reuse responses for next run.

**Reply by ChrisJBurns** [reply]
> also pulling official images or bitnami images are not counted to rate limit

Is this for certain? I just pulled an `alpine` image and it counted as a pull and I have 99 remaining. I can't see this anywhere on DockerHub rate limit pages :( 

> you should configure redis cache, so renovate can reuse responses for next run.

Ahh this may help a lot. So if it detects that a new version is available, it pulls it and stores it in the cache? Then next run, if it detects that there is a new version available and it recognises it already has it in the cache, it just uses the one in the cache instead of re-pulling it from DockerHub?

**Reply by rarkins** [reply]
I don't think that "library" images are free to pull. But Docker has a program where non-library publishers can pay Docker money so that users can pull the publisher's images for free. That might be the meaning of the "VERIFIED PUBLISHER" badge you see next to bitnami, hashicorp, etc.

**Reply by ChrisJBurns** [reply]
Gotcha! I've just pulled a Hashicorp image and can confirm it doesn't count towards the rate limit. On the Redis point, I've configured Redis for Renovate, although when I re-run it against the same repo, I'm still seeing the rate limits decrease each time. Not sure if there are circumstances where it still pulls?

For additional information, I'm also using persistent caching also. Lastly, when i removed the Redis URL from Renovate (just to see the difference in rate limit decreases) and I'm still seeing a similar rate of decrease from our remaining pulls and the Renovate run doesn't get any faster. So Redis doesn't seem to make much of a difference for us at a first look :(
